server/github/client.go

```diff
@@ -0,0 +1,89 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/Thesohan/weaveGitHubSearchService/server/constants"
+	httpClient "github.com/Thesohan/weaveGitHubSearchService/server/http_client"
+	"github.com/Thesohan/weaveGitHubSearchService/server/structures"
+)
+
+// GitHubAPIURL is the base URL for GitHub search API.
+const GitHubAPIURL = "https://api.github.com/search/code"
+
+var (
+	once      sync.Once
+	singleton IGithubClient
+)
+
+type githubClient struct {
+	client  *http.Client
+	baseURL string
+	headers map[string]string
+}
+
+// IGithubClient interface for API requests.
+type IGithubClient interface {
```

Comment:
Idomatic go does not use an `I` prefix. Instead follow the [effective go](https://go.dev/doc/effective_go#interface-names) advice. Also don't make anything public unless it absolutely needs to be public.

```suggestion
// codeSearcher interface for API requests.
type codeSearcher interface {
```